includes/abstracts/llms.abstract.notification.view.php

```diff
@@ -617,6 +632,28 @@ private function get_merged_string( $string ) {
 
 	}
 
+	/**
+	 * Retrieve merge codes used in a given string.
+	 *
+	 * @since [version]
+	 *
+	 * @param string $string Text string whereto look for merge codes.
+	 * @return array Returns a list of merge codes actually used in the passed string.
+	 */
+	private function get_used_merge_codes( $string ) {
+
+		return array_keys(
```

Comment:
I'll probably never need to debug this method, but in the future, please keep in mind that it is difficult to step debug complex functions when they are not assigned to a variable. I can't see the result of this series of function calls because it is immediately returned. Plus, this method uses the return value directly in a `foreach` loop, so I can't see it there either.

Reply:
I see, I will try to make stuff more readable in the future but I really think this one is not complex to read.

Reply:
It's difficult to use Xdebug to step through live running code and see the contents of variables. The readability is OK.

Reply:
Oh I see.
Is it better to declare several _useless_ variables for "maybe troubleshooting" purpose only?
I do understand your concern, but as long as something is _readable_ (which I understand is personal) we can always unpack when needed.

The following code would be ok for Xdebug?
```php
	private function get_used_merge_codes( $string ) {
		
		$merge_codes      = $this->get_merge_codes();
		$used_merge_codes = array_filter(
			$merge_codes,
			function ( $code ) use ( $string ) {
				return false !== strpos( $string, $code );
			},
			ARRAY_FILTER_USE_KEY
		);
		$used_merge_codes = array_keys( $used_merge_codes );

		return $used_merge_codes;

	}
```

Reply:
Yes, that code would be easy to step through and see the data as it flows through functions.

Assigning the result of a function to a variable before returning the value is most useful when I'm not sure what the result will be, such as when `apply_filters()` is used. Sometimes it's faster for me to step debug through the code than to unpack multiple function calls in my head.

Reply:
Do you think using variables when not needed would have any "micro" performance impact though?

Reply:
Yes, there is a performance impact. From what I know about the internal Zend engine, it's probably **very** small. When a function returns data, the data is already taking up memory. To assign it to a variable, Zend allocates a small block of memory for the name, type, and stores the pointer to the memory that holds the data.

I woke up too dang early this morning, so I think just for fun I'll benchmark it.

Reply:
@eri-trabiccolo, my benchmark results were surprising.

I wrote two functions that both call `str_repeat()`. One returns it directly, the other assigns it to a variable and then returns the variable. Each function is called 10,000 times. The number of times that the string should be repeated increments with the number of times the function is called.

**Sometimes function 1 is faster than function 2. The other times it is reversed.** The difference can be as great as 9 milliseconds.

It seems that there is no performance impact of assigning a function's result to a variable and then returning the variable.

```php
function return_function_result( $times ) {
	return str_repeat( 'A', $times );
}
function return_function_variable( $times ) {
	$result = str_repeat( 'A', $times );
	return $result;
}

$loopLimit = 10000;

$start = microtime(true);
for ( $i = 0; $i < $loopLimit; $i++ ) {
	$string = return_function_result( $i );
}
$stop = microtime(true);
$return_function_result_duration = $stop - $start;
echo "<pre>return_function_result()   = " . sprintf('%.2f', $return_function_result_duration * 1000) . "\n";

$start = microtime(true);
for ( $i = 0; $i < $loopLimit; $i++ ) {
	$string = return_function_variable( $i );
}
$stop = microtime(true);
$return_function_variable_duration = $stop - $start;
echo "return_function_variable() = " . sprintf('%.2f', $return_function_variable_duration * 1000) . "\n";

$difference = $return_function_variable_duration - $return_function_result_duration;
echo "                DIFFERENCE = " . sprintf('%.2f', $difference * 1000 ) . "</pre>";
```

Reply:
I see,
well I think in this case performance should be referred to memory usage maybe rather than execution speed, and in a little bit more complex functions, where there are several declared variables, I mean something more similar to the method I wrote and the method Xdebug would like the best. :)

Reply:
The memory used by a variable is basically the length of the variable's string name plus a pointer to the  [zval structure](https://www.phpinternalsbook.com/php7/zvals/basic_structure.html#the-zval-struct). The size of the zval is the same whether assigned to a variable or used/returned immediately.

Unless our methods are assigning a large number of variables, or the variable names are really long, the memory used and the execution time to allocate the memory are insignificant. When the method returns, the memory used by variables is reclaimed by PHP's garbage collector.

For a deeper dive, see https://github.com/php/php-src/blob/master/main/php_variables.c.

Reply:
Thanks.
> When the method returns, the memory used by variables is reclaimed by PHP's garbage collector.

Which means that the memory used by variables is retained at least until the method returns.
Would be interesting to know what happens instead to the memory used by the zval structure of a function used immediately, how long it lasts I mean.

Reply:
When returning from a function or method, the reference counter of zvals in that scope is decremented if they aren't used in the return. When a zval's reference counter reaches 0, the garage collector is allowed to reclaim the memory.

Reply:
So we're saying that:
```php
		return array_keys(
			array_filter(
				$this->get_merge_codes(),
				function ( $code ) use ( $string ) {
					return false !== strpos( $string, $code );
				},
				ARRAY_FILTER_USE_KEY
			)
		);
```
is better than:
```php
		$merge_codes      = $this->get_merge_codes();
		$used_merge_codes = array_filter(
			$merge_codes,
			function ( $code ) use ( $string ) {
				return false !== strpos( $string, $code );
			},
			ARRAY_FILTER_USE_KEY
		);
		$used_merge_codes = array_keys( $used_merge_codes );

		return $used_merge_codes;
```

in that _very_ respect?

Reply:
Yes. Assigning the results of function calls to locally scoped variables allows the value to be examined during [step debugging](https://xdebug.org/docs/step_debug). The memory used is insignificant and is reclaimed if needed.

Using function calls in arguments to other function calls is hard to debug. It can also be difficult to read if there are many nested function calls.